### DISCUSSION TITLE
[2023-06-12] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#7015] [Weekly Demo Day]
I finished the implementation of the new FRP engine and created a [PR for it](https://github.com/enso-org/enso/pull/6944) that is now being reviewed by Adam. The PR does not have nice description and the code has some places that should be refactored (e.g. a very long file with implementation), but I'm facing some unexpected problems that I'm trying to resolve. The biggest one is that refactoring this file to different files causes a 20-30% performance regression (only refactoring, no logic change) even when compiled with `LTO="fat"`. 

The new FRP implementation is not yet integrated with the product (this will be a separate PR), however, it's worth mentioning a few of its benefits:
- Creation of FRP nodes is a lot faster (50-100x) the old implementation.
- Sending FRP events with simple, copyable data is up to 2x faster than the old implementation.
- Sending complex data is way faster (depending on the data), as now, the FRP sends the data via references. E.g. sending strings makes it 20-30x faster.
- The FRP Network owns now a model, giving direct access to mutable model in some functions, such as `map`. 
- Definition of the FRP nodes is extremely simple, in most cases an FRP node definition is 2-3 lines of code, including the signature. In comparison, definition of a single node in the old implementation was always 50-100 lines long. 
- There is a new API for defining FRP, that does not require the `frp::extend!` macro anymore! For example, you can now do:

   ```rust
        let net = Network::new();
        let n1 = net.source::<usize>();
        let n2 = n1.map_(|t| t + 1);
        let mut prev = n2;
        for _ in 0..8 {
            let next = prev.map_(|t| t + 1);
            prev = next;
        }
   ```

- In order to deliver the PR, I needed to improve our generics library, which now is nicely implemented and allows for pretty sophisticated operations. Basically, it allows you to work in a generic fashion with fields of **ANY DATA STRUCTURE**, including your own data structures, even if the fields are of different types (it even allows iteration over the fields and mapping them with type-dependent functions). For example, for tuples, the following are possible:

   ```rust
        let tuple = (1, 1.0, vec![2, 3]);
        assert_eq!(tuple.field_count(), 3);

        let tuple = (1, 1.0, vec![2, 3]);
        assert_eq!(tuple.as_ref_fields(), (&1, &1.0, &vec![2, 3]));

        let mut tuple = (1, 1.0, vec![2, 3]);
        *tuple.first_field_mut() += 2;
        assert_eq!(tuple, (3, 1.0, vec![2, 3]));

        let mut tuple = (1, 1.0, vec![2, 3]);
        *tuple.field_at_mut::<0>() += 2;
        *tuple.field_at_mut::<1>() += 3.0;
        tuple.field_at_mut::<2>().push(4);
        assert_eq!(tuple, (3, 4.0, vec![2, 3, 4]));

        let tuple = (1, 1.0, vec![2, 3]);
        assert_eq!(tuple.push_last_field(7), (1, 1.0, vec![2, 3], 7));

        let tuple = (1, 1.0, vec![2, 3]);
        assert_eq!(tuple.pop_last_field(), (vec![2, 3], (1, 1.0)));

        let tuple = (1_u16, 2_u16, 3_u16);
        assert_eq!(tuple.map_fields_into::<f32>(), (1.0, 2.0, 3.0));

        let mut tuple = (1, 2, 3);
        tuple.field_iter_mut(|t| *t += 1);
        assert_eq!(tuple, (2, 3, 4));
   ```

**Comment by vitvakatu** [comment]
- Fixed the issue with an invalid cursor position after ctrl-clicking the node (https://github.com/enso-org/enso/issues/6702): the PR is on QA.
- Opened PR with CSS updates in the documentation panel, bringing multiple fixes and improvements https://github.com/enso-org/enso/pull/6978
- We now display an error message when receiving an `execution failed` message from the engine (the PR got finally merged https://github.com/enso-org/enso/pull/6918)

https://github.com/enso-org/enso/assets/6566674/e4cef3d1-7a13-4d3b-b11d-f5ee2c05ca82

**Comment by MichaelMauderer** [comment]
* Finished up visualization fixes. 
* Investigating #6908, which turns out to be another _interesting_ asynchronous event order issue that can be made to go away by adding some logging in the right/wrong place.

**Comment by farmaazon** [comment]
**Fixed [problems with typing text in HTML inputs](#6981)** (in new dashboard or inside visualizations). The bug was caused by preventing default on keyboard events. Now we do prevent, but only on some shortcuts (which were a reason we had added the preventing in the first place).

On this occasion I also tried to capture events in div created by EnsoGL, not on the entire document, but the focus was hard to manage in the HTML (usually the user needs to click on canvas on various occasions to have the shortcut working), so I gave up.

**Trace down and fixed [last leaks of Application](#6911) after calling drop on WASM**. This fixed the multiple warnings after opening another project in the new dashboard, which indicated that some rendering of the previous project was still there. 

The main cause of leaks was - as usual - the wrong object being captured in FRP nodes (the object which owns this FRP network). The new FRP system of @wdanilo should make such bugs much harder to introduce. Additionally, I discovered that our ["garbage collector"](https://github.com/enso-org/enso/blob/develop/lib/rust/ensogl/core/src/display/garbage.rs) may also cause some leaks, but I'm not sure if it actually happened.

**Comment by kazcw** [comment]
- My next performance PR is ready for review #6954. I fixed visualization attachment logic, sped up deserializing messages 6x, and made edge recoloring quick. We now do a lot less frontend CPU work when handling any changes that affect large parts of the graph, including the case of #6674:

|before|after|
|---|---|
|![vokoscreenNG-2023-06-09_08-49-46.webm](https://github.com/enso-org/enso/assets/1047859/a0048b32-4906-41cd-8899-6e2543ef6942)|![vokoscreenNG-2023-06-09_08-50-54.webm](https://github.com/enso-org/enso/assets/1047859/fef81512-ad89-4418-ae10-d54de94d96ea)|

- I started on #6865. Due to a bug in dropdowns, the `GridView` isn't lazy. I'll fix it today.

**Comment by mwu-tow** [comment]
Last week I was finishing work on Undo/Redo improvements (https://github.com/enso-org/enso/pull/6950). Applying CR feedback, debugging, and fixing the issues found during the QA (including one old issue).
I took over and started to work on https://github.com/enso-org/enso/pull/6582 to reenable remote logging.

**Comment by Frizi** [comment]
- continuing work on refactoring edges. #6772 I have effectively rewritten the edge connection handling logic within graph editor, in order to allow edge views to have separate state from connections visible by graph controller. Still finishing integrating those changes with node views.
- Discussed and agreed on language server protocol changes needed for more complete method argument resolution at IDE side.  https://github.com/enso-org/enso/issues/6957

**Comment by JaroslavTulach** [comment]
Engine-ers focus on non-performance related things - mostly robustness:
- #6936
- #6821
- #6900
- #6897
- Deadlocks - #6841
